earth2studio/data/cmip6.py

```diff
@@ -0,0 +1,420 @@
+            grid_shape = (len(lat_1d), len(lon_1d))
+            da_dims_xy = ("lat", "lon")
+            coord_dict: dict[str, object] = {
+                "lat": lat_1d,
+                "lon": lon_1d,
+            }
+        # Curvilinear ocean/ice grid → 2-D latitude/longitude coordinates
+        elif {"latitude", "longitude"} <= set(ds.coords):
+            lat2d = ds["latitude"].values
+            lon2d = ds["longitude"].values
+            # Use underlying dimension names (typically j,i or y,x)
+            y_dim, x_dim = ds["latitude"].dims
+            grid_shape = lat2d.shape  # (ny, nx)
+            da_dims_xy = (y_dim, x_dim)
+            coord_dict = {
+                y_dim: ds[y_dim],
+                x_dim: ds[x_dim],
+                "_lat": (da_dims_xy, lat2d),
+                "_lon": (da_dims_xy, lon2d),
+            }
+        else:  # pragma: no cover
+            raise ValueError(
+                "Unable to determine horizontal coordinates for CMIP6 dataset – expected 'lat/lon' or 'latitude/longitude'."
+                f"\nDataset: {ds}"
+            )
+
+        # Find the nearest available times in the first dataset.
+        time = self._convert_times_to_cftime(time, ds.time.dt.calendar)
+        ds_nearest = ds.sel(time=time, method="nearest")  # type: ignore[arg-type]
+        selected_times = ds_nearest.time.values.tolist()
+        if any(t_req != t_sel for t_req, t_sel in zip(time, selected_times)):
+            warnings.warn(
+                "One or more requested timestamps were not found exactly in the CMIP6 dataset; "
+                "nearest available snapshots have been substituted.",
+                UserWarning,
+            )
+        time = selected_times
+
+        # Subset time; rely on xarray's KeyError if a timestamp is missing
+        for var_name, ds_var in dsd.items():
+            dsd[var_name] = ds_var.sel(time=time)  # type: ignore[arg-type]
+
+        # Make data array
+        da = xr.DataArray(
+            data=np.empty((len(time), len(variable), *grid_shape), dtype=np.float32),
+            dims=["time", "variable", *da_dims_xy],
+            coords={
+                "time": self._convert_times_to_datetime(time),
+                "variable": variable,
+                **coord_dict,
+            },
+        )
+
+        # Populate the array
+        for var_idx, var in enumerate(variable):
+
+            # Get variable, level, and modifier
+            EntryT = tuple[
+                tuple[str, int], Callable[[np.ndarray], np.ndarray]
+            ]  # NOTE: I couldn't figure out hot to fix the ruff error on this so just added this type hint
+            cmip6_entry = cast(EntryT, CMIP6Lexicon.get_item(var))
+            (cmip6_var, level), modifier = cmip6_entry
+
+            # Get data array
+            ds_var = dsd[cmip6_var]
+            data_arr = ds_var[cmip6_var]  # DataArray inside the dataset
+
+            # Select pressure level if needed
+            if level != -1:
```

Comment:
Is this if statement tested? See a large gap in the tests here. Wanna make sure the level testing is working as expected.